Implement Change Password Feature (Issue #279)

iamericfletcher opened the pull request:
### Changes made

Implemented new change password feature. 

- Created new `PwChangeModal` and added it to `src/routes/profile/modals/PwChangeModal.svelte`
- Created `changeUserPassword` in `src/lib/util/api.ts`
- Created `POST("/change_password")` with `addUserRoutes` in `server/routes.go` - so `/user/change_password`
- Created `userChangePassword` in `server/user.go` with `UserPasswordUpdateRequest` struct in `server/auth.go`

Thoughts:

- Could probable create a `FormList` similar to `SettingsList` (`src/lib/settings/SettingsList.svelte`)

```
  .form-input-container {
    display: flex;
    flex-flow: column;
    gap: 20px;
    margin: 0 15px;
  }

  .form-input-container > button {
    margin-left: auto;
    width: max-content;
  }
```

iamericfletcher committed:
Added beginning stages of password change functionality in user profile

A new interface ChangePasswordForm has been created to handle password change form data.

The 'PwChangeModal.svelte' modal allows the user to input current and new passwords for a change password request.

The user profile page has been updated to include a "Change Password" button which triggers the change password modal.

iamericfletcher committed:
Refactoring password change modal form

iamericfletcher committed:
Refactoring password button in PwChangeModal to be in line with button in SonarrModal

iamericfletcher committed:
Improve form validation in PwChangeModal.svelte

Updated form validation to check for missing inputs and ensure consistency between new password and re-entered new password. Enhanced logging for better status tracking and error reporting during form validation process.

iamericfletcher committed:
Implement password change feature

A new feature has been added that allows users to change their password. This includes updates in several different files to accommodate this new functionality. Changes include adding a new API endpoint, creating a new function to handle the password change request, and removing the username field from the ChangePasswordForm interface.

iamericfletcher committed:
Refactor user password change functionality

The password change function now involves first querying user data from the database, before checking the validity of their old password. Logic has been updated on both frontend and backend code. Errors have been added for handling instances where password comparison, user retrieval or other processes fail.

iamericfletcher committed:
Refine password change error handling and reporting

The commit delivers improved error handling and debugging information for password change functionality. Error messages now provide more detailed context on failures. Additionally, new password hashing has been added in the password change process. The user interface has also been updated to reflect specific error messages returned from the API.

iamericfletcher committed:
Refactor password change feature and update debug messages

Removed a substantial block of commented-out code related to password changing in the `user.go` file, which simplifies the structure and readability of the code. Improved clarity and specificity of console debug/feedback messages throughout the frontend, ensuring easier debugging. Commit also includes minor code clean up in some files.

iamericfletcher committed:
Remove Form.svelte and related import

Form.svelte was deleted due to not being used or needed. An associated import line in PwChangeModal.svelte has also been removed.

Additionally, setting parameter DEBUG was removed from server/config.go.

IRHM commented on server/user.go:
```diff
@@ -164,3 +164,46 @@ func uploadUserAvatar(c *gin.Context, db *gorm.DB, userId uint) (Image, error) {
 	return img, nil
 }
+
+func userChangePassword(db *gorm.DB, pwds UserPasswordUpdateRequest, userId uint) error {
```
This may fit better in `auth.go` (This function is more related to auth than what else is in `user.go`). The `user.go` file is more for editing settings or finding other users.

iamericfletcher replied:
I agree! I was on the fence about this myself and am glad you mentioned this :)

IRHM commented on server/user.go:
```diff
@@ -164,3 +164,46 @@ func uploadUserAvatar(c *gin.Context, db *gorm.DB, userId uint) (Image, error) {
+	if res.Error != nil {
+		slog.Error("userChangePassword failed - failed to retrieve user from database", "user_id", userId, "error", res.Error)
+		return errors.New("failed to retrieve user")
+	} else {
```
The above `if` returns, so it acts like a guard. We could help with readability by removing the `else`.

IRHM commented on server/user.go:
```diff
@@ -164,3 +164,46 @@ func uploadUserAvatar(c *gin.Context, db *gorm.DB, userId uint) (Image, error) {
+		if err != nil {
+			slog.Error("userChangePassword failed - failed to hash new password", "user_id", userId, "error", err)
+			return errors.New("failed to hash new password")
+		} else {
```
Same as above. We could help with readability by removing the `else`.

IRHM commented on server/user.go:
```diff
@@ -164,3 +164,46 @@ func uploadUserAvatar(c *gin.Context, db *gorm.DB, userId uint) (Image, error) {
+		}
+		slog.Debug("userChangePassword hash for current password matches hash in the database", "user_id", userId)
+		slog.Debug("userChangePassword hashing new password", "user_id", userId)
+		hash, err := hashPassword(pwds.NewPassword, &ArgonParams{
```
This is up to you if you want to. Since the `ArgonParams` must be the same here as in the `register` function, we could move them both out to a `func` (in `auth.go`):

```
func GetPassArgonParams() *ArgonParams {
	return &ArgonParams{
		memory:      64 * 1024,
		iterations:  3,
		parallelism: 2,
		saltLength:  16,
		keyLength:   32,
	}
}
```

I'm leaning towards a func (instead of a var) so we can rest easy knowing the parameters can't be changed (for everything).

iamericfletcher replied:
Totally agree in `ArgonParams` being a `func` instead of `var`!

IRHM commented on src/lib/util/api.ts:
```diff
@@ -170,6 +170,23 @@ export function updateUserSetting<K extends keyof UserSettings>(
+    .catch((err) => {
+      const errMsg = err.response.data.error ? err.response.data.error : "Couldn't Change Password";
+      console.error("Change Password Form - Failed to change password on the server", err);
+      notify({ id: nid, type: "error", text: errMsg });
```
Instead of using `notify` here, we could pass the error to the `done` callback. In the modal, we could use that to set `error` to show an error box at the top.

I think this should make it easier for users, so all errors show up in the same place.

IRHM commented on src/routes/(app)/profile/modals/PwChangeModal.svelte:
```diff
@@ -0,0 +1,159 @@
+      const fd = new FormData(ev.target! as HTMLFormElement);
+      changepswd.currentPassword = fd.get("current-password") as string;
+      changepswd.newPassword = fd.get("new-password") as string;
+      changeUserPassword(changepswd.currentPassword, changepswd.newPassword, () => {
```
Related to the other comment, it could then be used like so:

```
changeUserPassword(changepswd.currentPassword, changepswd.newPassword, (err) => {
  if (err) {
    error = err;
    return;
  }
  console.log("Change Password Form - Status: Password changed successfully");
  onClose();
});
```

IRHM reviewed (changes requested):
Hi @iamericfletcher, this functions very well, thank you for working on it!

I added a few suggestions that I think could improve the changes. Let me know if you have any thoughts/questions about them.

iamericfletcher committed:
Move userChangePassword function to auth.go

iamericfletcher committed:
Move server endpoint to /auth/change_password and update corresponding frontend API call

iamericfletcher committed:
Removed unneeded `else` branches in `auth.go`

iamericfletcher committed:
Refactor ArgonParams to a function for `register` and `userChangePassword` in `auth.go`

iamericfletcher committed:
Update error handling in `changeUserPassword` and display of error messages in `pwChangeModal`

The 'changeUserPassword' callback is updated such that error messages are passed back to the calling function in `PwChangeModal` allowing errors to show up in the same place.

IRHM committed:
Move GetPassArgonParams, Remove DEBUG, add use of ternary operator

IRHM committed:
userChangePassword: Select only pass from db

IRHM reviewed (dismissed):
Works great! Thank you again!

I added a few small changes I thought weren't worth hassling you over. Mainly:

- `SELECT` only the password from database
- Use `?` where we set `errMsg` to avoid erroring if `err`, `response` or `data` are undefined.

IRHM committed:
Merge branch 'dev' into user-password-change-issue-279

iamericfletcher commented:
> Works great! Thank you again!
> 
> I added a few small changes I thought weren't worth hassling you over. Mainly:
> 
> * `SELECT` only the password from database
> * Use `?` where we set `errMsg` to avoid erroring if `err`, `response` or `data` are undefined.

You're very welcome! Learning a lot from you and this project, so thank you :)
